File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public List<String> solution(String[] players, String[] callings) {
        List<String> answer = new ArrayList<>();
        
        HashMap<String, Integer> rank_p = new HashMap<>();
        HashMap<Integer, String> rank_c = new HashMap<>();
        int cnt = 1;
        for(String player : players){
            rank_p.put(player, cnt);
            rank_c.put(cnt, player);
            cnt++;
        }
        
        
        
        for(String caller : callings){
            String chaser = caller;
            Integer chaserRank = rank_p.get(chaser);
            
            Integer headerRank = chaserRank - 1;
            String header = rank_c.get(headerRank);
            
            rank_p.remove(chaser);
            rank_p.remove(header);
            
            rank_c.remove(chaserRank);
            rank_c.remove(headerRank);
            
            rank_c.put(headerRank, chaser);
            rank_p.put(chaser, headerRank);
            
            rank_c.put(chaserRank, header);
            rank_p.put(header, chaserRank);
            
        }
        
        for(int i=1; i<cnt; i++) 
            answer.add(rank_c.get(i));
        
        return answer;
    }
}

File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    answer = []
    
    rank_p = {}
    rank_c = {}
    cnt = 1
    for p in players:
        rank_p[p] = cnt
        rank_c[cnt] = p
        cnt += 1
    
    for cal in callings:
        # 불린사람
        chaser = cal # 불린 사람 이름
        chaser_rank = rank_p[chaser] # 불린 사람의 현재 순위
        
        header_rank = chaser_rank - 1 # 앞선 사람의 순위
        header =  rank_c[chaser_rank-1] # 앞선 사람의 이름
        
        del rank_p[chaser]
        del rank_p[header]
        
        del rank_c[chaser_rank]
        del rank_c[header_rank]
        
        # 뒤의 사람은 한 칸 앞으로 간다.
        rank_c[header_rank] = chaser
        rank_p[chaser] = header_rank
        
        # 앞의 사람은 한 칸 뒤로 간다.
        rank_c[chaser_rank] = header
        rank_p[header] = chaser_rank
    for i in range(1, cnt):
        answer.append(rank_c[i])
        
    return answer